python/dasclient/das_agol_tools.py

```diff
@@ -0,0 +1,531 @@
+        """
+        Replaces the attachments of Esri features with the attachments described
+        by the event_files parameter.
+
+        :param esri_layer: The AGO layer to work with.  Features are loaded
+            which match the ER report numbers specified within the event_files
+            paramter
+        :param oldest_date: As far back to look for matching Esri features
+        :param event_files: Describes the files to upload.  This is a dictionary of the form:
+            {
+                er_report_number: {
+                    'url': URL to load the file from
+                    'filename': Filename to give the file
+                }
+            }
+        :return: None
+        """
+        existing_events = self._get_existing_esri_events(esri_layer, oldest_date)
+        tmpdir = tempfile.TemporaryDirectory()
+
+        for event in event_files.keys():
+            esri_object_id = existing_events[event][0]
+            existing_attachments = esri_layer.attachments.get_list(esri_object_id)
+
+            for existing_file in existing_attachments:
+                self.logger.info(f"Removing attachment {existing_file['name']} from feature {esri_object_id}")
+                esri_layer.attachments.delete(esri_object_id, existing_file['id'])
+
+            for file in event_files[event]:
+                self.logger.info(f"Adding attachment {file['filename']} from ER event {event} to Esri feature {esri_object_id}")
+                tmppath = tmpdir.name + "/" + file['filename']
+                result = self.das_client.get_file(file['url'])
+                open(tmppath, 'wb').write(result.content)
+                esri_layer.attachments.add(esri_object_id, tmppath)
+
+        tmpdir.cleanup()
+
+    def upsert_tracks_from_er(self, esri_layer, since):
+        """
+        Queries all EarthRanger subjects from the active ER connection, grabs
+        their tracks, and creates or updates polylines in AGO to match.  The
+        EarthRanger subject ID is stored in AGO as a parameter ER_ID on each
+        linestring.
+
+        Right now, this method is very brute force: It either adds a track or
+        replaces the whole track.  There are also not (yet) any parameters
+        around track length, date ranges, which subjects to include, etc.
+
+        :param esri_layer: The AGO layer to upsert
+        :param since: The start date/time of the track to load
+        :return: None
+        """
+        self._ensure_attributes_in_layer(esri_layer, [
+            {'name':'ER_ID', 'alias': 'ER ID', 'type': 'esriFieldTypeString'},
+            {'name':'SUBJECT_NAME', 'alias': 'Subject Name', 'type': 'esriFieldTypeString'}
+        ])
+
+        subjects = self.das_client.get_subjects()
+        existing_tracks = self._get_existing_esri_tracks(esri_layer)
+        features_to_add = []
+        features_to_update = []
+
+        if(since == None):
+            since = pytz.utc.localize(datetime.utcnow()) - timedelta(days=30)
```

Comment:
I settled on using only the python lib for getting a non-naive datetime of now: datetime.datetime.now(tz=datetime.timezone.utc)